### PR TITLE
Improve DecodableList when not in list

### PR DESCRIPTION
**Pull request by mlepage-google**
#### Problem
DecodableList doesn't behave well if you don't supply a reader
with a well-formed list.

#### Change overview
Add a ClearReader method, used by the constructor but also
reusable later, that initializes the internal reader to a
container-less state. Check for this container-less state in
ComputeSize and iterator methods such as Next.

#### Testing
Built and ran all-clusters-app on Linux with REPL to try null lists
as well as lists with items.

### CODE REVIEW

File: src/app/data-model/DecodableList.h [hunk 1]
```diff
@@ -148,13 +173,18 @@ class DecodableList
...
 CHIP_ERROR Decode(TLV::TLVReader & reader, DecodableList<X> & x)
 {
-    VerifyOrReturnError(reader.GetType() == TLV::kTLVType_Array, CHIP_ERROR_SCHEMA_MISMATCH);
-
-    TLV::TLVType type;
-
-    ReturnErrorOnFailure(reader.EnterContainer(type));
-    x.SetReader(reader);
-    ReturnErrorOnFailure(reader.ExitContainer(type));
+    if (reader.GetType() == TLV::kTLVType_Null)
```
Comment (bzbarsky-apple) [inline src/app/data-model/DecodableList.h:176]: What the spec says is:

1) Null and empty list must have the same semantics, _for a list that is nullable_.
2) Null is encoded as TLV null.
3) Empty list is encoded as empty TLV array.
4) Some lists are nullable, some are not.

There was some discussion about alternate encodings for empty non-nullable lists, but that's not where the spec landed in the end.

Nullable lists are represented in our C++ as `Nullable<DecodableList>` and you can decode into that and get the handling of TLV null.

Non-nullable lists are represented as just `DecodableList`, and a TLV null on the wire is not a valid encoding for those.  This change seems to want to allow TLV null on the wire for non-nullable lists...

Why are we making this change?

File: src/app/data-model/DecodableList.h [hunk 2]
```diff
@@ -134,12 +147,24 @@ class DecodableList
...
+     * actually decoding them. If there is no list (null valued list case) then
+     * the size is considered to be zero.
      */
-    CHIP_ERROR ComputeSize(size_t * size) const { return mReader.CountRemainingInContainer(size); }
+    CHIP_ERROR ComputeSize(size_t * size) const
+    {
+        if (mReader.GetContainerType() == TLV::kTLVType_NotSpecified)
+        {
+            *size = 0;
+            return CHIP_NO_ERROR;
```
Comment (bzbarsky-apple) [inline src/app/data-model/DecodableList.h:161]: That said, this part we might need for a default-initialized DecodableList to represent an empty list, and the same for the `Next` change....

File: src/app/data-model/DecodableList.h (hunk 1 above)
Comment (mlepage-google) [reply src/app/data-model/DecodableList.h:176]: Ah, good point.

File: src/app/data-model/DecodableList.h (hunk 2 above)
Comment (mlepage-google) [reply src/app/data-model/DecodableList.h:161]: Yeah this might be handy for the default-initialized DecodableList as having better behaviour. You can also reset a DecodableList to this (using ClearReader) if that's desirable.

File: src/app/data-model/DecodableList.h [hunk 3]
```diff
@@ -148,13 +173,18 @@ class DecodableList
 template <typename X>
 CHIP_ERROR Decode(TLV::TLVReader & reader, DecodableList<X> & x)
 {
-    VerifyOrReturnError(reader.GetType() == TLV::kTLVType_Array, CHIP_ERROR_SCHEMA_MISMATCH);
```
Comment (bzbarsky-apple) [inline src/app/data-model/DecodableList.h:151]: So this looks good to me if we undo the changes to Decode itself.  The other parts are needed to make the ClearReader() case actually behave right, and I really appreciate you catching that it didn't!

Comment (mlepage-google) [reply src/app/data-model/DecodableList.h:151]: Decode changes removed.

File: src/app/data-model/DecodableList.h (hunk 1 above)
Comment (mlepage-google) [reply src/app/data-model/DecodableList.h]: Decode changes removed. Other changes kept with adjusted documentation.

**Review by bzbarsky-apple — APPROVED** [review]

**Review by andy31415 — APPROVED** [review]
